Run __setstate__ when cloning modules

bzinodev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45858 Run __setstate__ when cloning modules**

When cloning a module that has __setstate__, __getstate__ methods.
We need to load these methods to initialize these modules.

Differential Revision: [D24116524](https://our.internmc.facebook.com/intern/diff/D24116524)

bzinodev committed:
Run __setstate__ when cloning modules

When cloning a module that has __setstate__, __getstate__ methods.
We need to load these methods to initialize these modules.

[ghstack-poisoned]

supriyar commented:
@bzinodev can you also make the same change in the quantization code? We have a clone implementation here https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/quantization/insert_observers.cpp#L96

eellison reviewed (commented):
Looks good to me, but I would like another set of eyes from something with more knowledge of setstate / getstate - cc @suo, @jamesr66a

bzinodev committed:
Update on "Run __setstate__ when cloning modules"


When cloning a module that has __setstate__, __getstate__ methods.
We need to load these methods to initialize these modules.

Differential Revision: [D24116524](https://our.internmc.facebook.com/intern/diff/D24116524)

[ghstack-poisoned]

bzinodev committed:
Update on "Run __setstate__ when cloning modules"


When cloning a module that has __setstate__, __getstate__ methods.
We need to load these methods to initialize these modules.

Differential Revision: [D24116524](https://our.internmc.facebook.com/intern/diff/D24116524)

[ghstack-poisoned]

suo reviewed (approved):
This looks fine. If getstate is defined it should be called on a deepcopy. It appears that `clone` is a deepcopy.

Some followups that could be good (doesn't have to be in this PR, just recording):
1. deepcopy should do the same thing if it doesn't already.
2. Can we please rename `clone` to something actually meaningful? I had to read the implementation to see if it was a copy vs. deepcopy, since literally just says "clone" in the method documentation. Maybe, deepcopy_with_type cc @jerry39213gh
